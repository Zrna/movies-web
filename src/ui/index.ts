import * as styles from './styles';
import * as theme from './theme';

// export theme, enums, styles
export { styles, theme };
