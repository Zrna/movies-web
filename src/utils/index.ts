export * from './async';
export * from './error';
export * as validator from './validation';
