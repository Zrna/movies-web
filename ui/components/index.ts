export * from './Box';
export * from './Button';
export * from './Checkbox';
export * from './Divider';
export * from './ErrorMessage';
export * from './FlexLayout';
export * from './Icon';
export * from './Image';
export * from './LoadingSpinner';
export * from './Modal';
export * from './Only';
export * from './Search';
export * from './Text';
export * from './Textarea';
export * from './TextInput';
export * from './Toast';
