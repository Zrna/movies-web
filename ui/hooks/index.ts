export * from './dom';
export * from './modal';
