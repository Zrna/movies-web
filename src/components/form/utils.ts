import { FieldMetaState } from 'react-final-form';

import { TextInputProps } from '~/ui';

export interface FormFieldProps extends Omit<TextInputProps, 'value' | 'onChange'> {
  name: string;
  type?: string;
  label?: string;
  validate?: any;
  value?: null;
  onChange?(): null;
}

export function getFieldError(meta: FieldMetaState<any>) {
  if (meta.touched && meta.error) {
    return meta.error;
  }

  return meta.submitError;
}
