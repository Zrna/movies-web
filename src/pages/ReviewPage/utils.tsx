import { theme } from '~/ui';

export const reactionsMap: { [key: number]: { icon: theme.Icon; text: string } } = {
  0: { icon: 'reaction0', text: 'No comment' },
  1: { icon: 'reaction1', text: "I wish I hadn't watched" },
  2: { icon: 'reaction2', text: 'Boring' },
  3: { icon: 'reaction3', text: 'Not great, not terrible' },
  4: { icon: 'reaction4', text: "It's great" },
  5: { icon: 'reaction5', text: 'Loving it!' },
};
