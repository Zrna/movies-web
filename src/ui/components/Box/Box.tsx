import { forwardRef, useRef } from 'react';
import useClickAway from 'react-use/lib/useClickAway';
import { Box as ThemeBox, BoxProps as ThemeBoxProps, SxStyleProp } from 'theme-ui';

import { styles, theme } from '~/ui';

export interface BoxProps extends ThemeBoxProps {
  color?: theme.Color;
  isDisabled?: boolean;
  type?: string;
  value?: any;
  target?: string;
  sx?: SxStyleProp;
  onClick?: (event?: any) => void;
  onOutsideClick?(event: Event): void;
}

export const Box: React.FC<BoxProps> = forwardRef<any, any>(
  ({ isDisabled = false, sx, onClick, onOutsideClick, ...rest }, ref: any) => {
    const clickableStyle = isDisabled || onClick ? styles.interactions.clickable : {};

    const innerRef = useRef(ref);

    useClickAway(innerRef, () => {
      if (onOutsideClick) {
        onOutsideClick();
      }
    });

    return (
      <ThemeBox
        ref={innerRef}
        sx={{ ...clickableStyle, ...sx }}
        onClick={!isDisabled && onClick ? onClick : undefined}
        {...{ disabled: isDisabled, ...rest }}
      />
    );
  },
);
