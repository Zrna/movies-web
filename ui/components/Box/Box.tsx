import { forwardRef, useRef } from 'react';
import useClickAway from 'react-use/lib/useClickAway';
import { Box as ThemeBox, BoxProps as ThemeBoxProps, ThemeUIStyleObject } from 'theme-ui';

import { styles, theme } from '~/ui';

export interface BoxProps extends ThemeBoxProps {
  color?: theme.Color;
  isDisabled?: boolean;
  type?: string;
  value?: any;
  target?: string;
  sx?: ThemeUIStyleObject;
  onClick?: (event?: any) => void;
  onOutsideClick?(event?: Event): void;
  'data-testid'?: string;
}

export const Box = forwardRef<any, BoxProps>(
  ({ isDisabled = false, sx, onClick, onOutsideClick, 'data-testid': dataTestId, ...rest }, ref: any) => {
    const clickableStyle = isDisabled || onClick ? styles.interactions.clickable : {};

    const innerRef = useRef(ref);
    if (onOutsideClick) {
      useClickAway(innerRef, onOutsideClick);
    }

    return (
      <ThemeBox
        data-testid={dataTestId}
        ref={onOutsideClick ? innerRef : ref}
        sx={{ ...clickableStyle, ...sx }}
        onClick={!isDisabled && onClick ? onClick : undefined}
        {...{ disabled: isDisabled, ...rest }}
      />
    );
  },
);
