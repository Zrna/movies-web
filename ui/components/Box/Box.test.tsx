import { screen } from '@testing-library/react';

import { theme } from '~/ui';
import { expectTextVariant, getThemeColorString, renderWithTheme, userEvent } from '~/utils/tests';

import { Box } from './Box';

describe('Box', () => {
  it('should render', () => {
    renderWithTheme(
      <Box as="span" bg="dark" color="white" data-testid="box" isDisabled mx="auto" px={2} py={4} onClick={console.log}>
        <p>
          some <strong>strong</strong>
          content
        </p>
      </Box>,
    );

    expect(screen.getByTestId('box')).toMatchSnapshot();
  });

  describe('Margins', () => {
    it('should apply proper margins', () => {
      renderWithTheme(
        <Box data-testid="box1" m={2}>
          Box 1
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box2" m="auto">
          Box 2
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box3" m={-4}>
          Box 3
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box4" m="10vw">
          Box 4
        </Box>,
      );

      expect(screen.getByTestId('box1')).toHaveStyle({ margin: '8px' });
      expect(screen.getByTestId('box2')).toHaveStyle({ margin: 'auto' });
      expect(screen.getByTestId('box3')).toHaveStyle({ margin: '-16px' });
      expect(screen.getByTestId('box4')).toHaveStyle({ margin: '10vw' });
    });

    it('should apply proper horizontal and vertical margins', () => {
      renderWithTheme(
        <Box data-testid="box" mx={4} my={2}>
          Box
        </Box>,
      );

      expect(screen.getByTestId('box')).toHaveStyle({
        marginLeft: '16px',
        marginRight: '16px',
        marginTop: '8px',
        marginBottom: '8px',
      });
    });

    it('should apply proper directional margins', () => {
      renderWithTheme(
        <Box data-testid="box" mb={4} ml={1} mr={2} mt={3}>
          Box
        </Box>,
      );

      expect(screen.getByTestId('box')).toHaveStyle({
        marginLeft: '4px',
        marginRight: '8px',
        marginTop: '12px',
        marginBottom: '16px',
      });
    });

    it('should overwrite margins correctly', () => {
      renderWithTheme(
        <Box data-testid="box1" mt={4} my={2}>
          Box 1
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box2" ml={2} mx={4}>
          Box 2
        </Box>,
      );

      expect(screen.getByTestId('box1')).toHaveStyle({
        marginBottom: '8px',
        marginTop: '8px',
      });
      expect(screen.getByTestId('box2')).toHaveStyle({
        marginLeft: '16px',
        marginRight: '16px',
      });
    });
  });

  describe('paddings', () => {
    it('should apply proper padding', () => {
      renderWithTheme(
        <Box data-testid="box1" p={2}>
          Box 1
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box2" p="10vw">
          Box 2
        </Box>,
      );

      expect(screen.getByTestId('box1')).toHaveStyle({ padding: '8px' });
      expect(screen.getByTestId('box2')).toHaveStyle({ padding: '10vw' });
    });

    it('should apply proper horizontal and vertical padding', () => {
      renderWithTheme(
        <Box data-testid="box" px={4} py={2}>
          Box
        </Box>,
      );

      expect(screen.getByTestId('box')).toHaveStyle({
        paddingLeft: '16px',
        paddingRight: '16px',
        paddingTop: '8px',
        paddingBottom: '8px',
      });
    });

    it('should apply proper directional paddings', () => {
      renderWithTheme(
        <Box data-testid="box" pb={4} pl={1} pr={2} pt={3}>
          Box
        </Box>,
      );

      expect(screen.getByTestId('box')).toHaveStyle({
        paddingLeft: '4px',
        paddingRight: '8px',
        paddingTop: '12px',
        paddingBottom: '16px',
      });
    });

    it('should overwrite paddings correctly', () => {
      renderWithTheme(
        <Box data-testid="box1" pt={4} py={2}>
          Box 1
        </Box>,
      );
      renderWithTheme(
        <Box data-testid="box2" pl={2} px={4}>
          Box 2
        </Box>,
      );

      expect(screen.getByTestId('box1')).toHaveStyle({
        paddingBottom: '8px',
        paddingTop: '8px',
      });
      expect(screen.getByTestId('box2')).toHaveStyle({
        paddingLeft: '16px',
        paddingRight: '16px',
      });
    });
  });

  describe('colors', () => {
    it('should apply proper colors', () => {
      renderWithTheme(
        <Box bg="dimmed" color="primary" data-testid="box" mt={2}>
          Box
        </Box>,
      );
      const element = screen.getByTestId('box');

      expect(element).toHaveStyleRule('background-color', getThemeColorString('dimmed'));
      expect(element).toHaveStyleRule('color', getThemeColorString('primary'));
    });
  });

  describe('sx', () => {
    it('should apply sx prop correctly', () => {
      const variant = 'paragraph-default';

      renderWithTheme(
        <Box
          data-testid="box"
          sx={{
            border: `2px solid ${theme.colors['white-alpha-25']}`,
            borderRadius: '20px',
            height: '72px',
            variant: `text.${variant}`,
          }}
        >
          Box
        </Box>,
      );

      expect(screen.getByTestId('box')).toHaveStyle({
        border: '2px solid rgba(255, 255, 255, 0.25)',
        borderRadius: '20px',
        height: '72px',
      });

      expectTextVariant(screen.getByTestId('box'), variant);
    });
  });

  describe('`as` props', () => {
    it('should correctly render passed element', () => {
      renderWithTheme(
        <Box as="h1" data-testid="box1">
          Box1
        </Box>,
      );
      renderWithTheme(
        <Box as="span" data-testid="box2">
          Box2
        </Box>,
      );
      renderWithTheme(
        <Box as="p" data-testid="box3">
          Box3
        </Box>,
      );
      renderWithTheme(
        <Box as="input" data-testid="box4" placeholder="Test placeholder" value="test" onChange={() => null} />,
      );

      const href = 'https://google.com';
      renderWithTheme(
        <Box as="a" data-testid="box5" {...{ href }}>
          Box5
        </Box>,
      );

      expect(screen.getByTestId('box1').tagName.toLowerCase()).toBe('h1');
      expect(screen.getByTestId('box2').tagName.toLowerCase()).toBe('span');
      expect(screen.getByTestId('box3').tagName.toLowerCase()).toBe('p');
      expect(screen.getByTestId('box4').tagName.toLowerCase()).toBe('input');
      expect(screen.getByTestId('box5').tagName.toLowerCase()).toBe('a');
      expect(screen.getByTestId('box5')).toHaveAttribute('href', href);
    });
  });

  describe('interaction', () => {
    it('should be clickable with pointer cursor', async () => {
      const user = userEvent.setup();
      const onClick = jest.fn();

      renderWithTheme(
        <Box data-testid="box" onClick={onClick}>
          Box
        </Box>,
      );

      const element = screen.getByTestId('box');
      expect(element).toHaveStyle({ cursor: 'pointer' });

      await user.click(element);
      await user.click(element);
      expect(onClick).toHaveBeenCalledTimes(2);
    });

    it('should call onOutsideClick when clicked outside', async () => {
      const handleOutsideClick = jest.fn();
      renderWithTheme(
        <Box data-testid="box" onOutsideClick={handleOutsideClick}>
          Click outside me
        </Box>,
      );

      await userEvent.click(document.body);
      expect(handleOutsideClick).toHaveBeenCalled();
    });

    it('should not call onOutsideClick when clicked inside', async () => {
      const handleOutsideClick = jest.fn();
      renderWithTheme(
        <Box data-testid="box" onOutsideClick={handleOutsideClick}>
          Click inside me
        </Box>,
      );

      await userEvent.click(screen.getByTestId('box'));
      expect(handleOutsideClick).not.toHaveBeenCalled();
    });
  });
});
