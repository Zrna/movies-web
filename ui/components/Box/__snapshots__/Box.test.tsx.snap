// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box should render 1`] = `
<span
  class="css-1l3a6k8-Box"
  data-testid="box"
  disabled=""
>
  <p>
    some 
    <strong>
      strong
    </strong>
    content
  </p>
</span>
`;
